include/unifex/async_resource.hpp

```diff
@@ -0,0 +1,423 @@
+  followed by clearing the range
+  - for an optional<T>, async_destroy(opt<T>) is a no-op on empty
+  - async_destroy(*opt) on non-empty, followed by .reset()
+  - for a unique_ptr<T>, async_destroy is async_destroy(*ptr) followed by
+  .reset(), unless it's nullptr, in which case no-op
+*/
+
+class _container_base {
+protected:
+  enum class state_t : char {
+    ALLOCATED,
+    DESTRUCT_SPAWNED,
+    CONSTRUCTED,
+    DESTROYING,
+  };
+
+  void (*destruct_this_resource_)(_container_base*) noexcept;
+  unifex::any_sender_of<> (*schedule_destruct_)(
+      _container_base*, _container_base*) noexcept;
+  unifex::any_sender_of<> (*destroy_this_resource_)(_container_base*) noexcept;
+  void (*deleter_)(_container_base*) noexcept;
+  // TODO migrate to v2/async_scope before wide adoption
+  UNIFEX_NO_UNIQUE_ADDRESS unifex::manual_lifetime<unifex::v1::async_scope>
+      scope_;
+  _container_base* children_;
+  _container_base* next_;
+  _container_base* parent_;
+  // means destruction should start
+  unifex::async_manual_reset_event destructionEvt_;
+  std::mutex mutex_;
+  std::atomic<state_t> state_;
+
+  unifex::task<void> close_child_scopes() noexcept {
+    auto* adopted = [&]() noexcept {
+      std::lock_guard lock{mutex_};
+      return std::exchange(children_, nullptr);
+    }();
+    {
+      for (auto* child = adopted; child != nullptr; child = child->next_) {
+        co_await child->close_scope();
+      }
+    }
+    {
+      std::lock_guard lock{mutex_};
+      // TODO repeat when children_ != nullptr
+      auto** tail = &children_;
+
+      while (*tail) {
+        tail = &(*tail)->next_;
+      }
+
+      *tail = adopted;
+    }
+  }
+
+  unifex::task<void> destroy_resource() noexcept {
+    co_await destroy_this_resource_(this);
+    for (auto* child = children_; child != nullptr; child = child->next_) {
+      co_await child->destroy_resource();
+    }
+  }
+
+  void destruct_container() noexcept {
+    if (state() == state_t::DESTROYING) {
+      destruct_this_resource_(this);
+      scope_.destruct();
+      _container_base* child = children_;
+      while (child != nullptr) {
+        auto next = child->next_;
+        child->destruct_container();
+        child = next;
+      }
+    }
+    deleter_(this);
+  }
+
+  void deregister_child(_container_base* child) noexcept {
+    std::lock_guard lock{mutex_};
+    for (_container_base** head = &children_; *head; head = &((*head)->next_)) {
+      if (*head == child) {
+        *head = (*head)->next_;
+        return;
+      }
+    }
+    UNIFEX_ASSERT(false);  // not found
+  }
+
+  auto destruct_impl() noexcept {
+    return unifex::let_value(destructionEvt_.async_wait(), [this]() noexcept {
+      bool just = true;
+      switch (state()) {
+        case state_t::ALLOCATED:
+          // scheduled destruct moved the state from ALLOCATED
+          UNIFEX_ASSERT(false);
+          break;
+        case state_t::CONSTRUCTED: just = false; break;
```

Comment:
I don't like whatever clang-format rule is putting all of this on one line :(

Reply:
#485